frontend/openchat-client/src/utils/cachePrimer.ts

```diff
@@ -0,0 +1,149 @@
+import type {
+    ChatEvent,
+    ChatSummary,
+    EventsResponse,
+    EventWrapper,
+    IndexRange
+} from "openchat-shared";
+import { compareChats, missingUserIds, userIdsFromEvents } from "openchat-shared";
+import { toRecord } from "./list";
+import { Poller } from "./poller";
+import type { OpenChatAgentWorker } from "../agentWorker";
+import { selectedChatId } from "../stores/chat";
+import { boolFromLS } from "../stores/localStorageSetting";
+import { messagesRead } from "../stores/markRead";
+import { userStore } from "../stores/user";
+import { get } from "svelte/store";
+
+export class CachePrimer {
+    private pending: Record<string, ChatSummary> = {};
+    private runner: Poller | undefined = undefined;
+    private selectedChatId: string | undefined = undefined;
+
+    constructor(private api: OpenChatAgentWorker) {
+        selectedChatId.subscribe((chatId) => {
+            if (chatId !== undefined) {
+                delete this.pending[chatId];
+            }
+            this.selectedChatId = chatId;
+        });
+        debug("initialized");
+    }
+
+    processChatUpdates(previous: ChatSummary[], next: ChatSummary[]): void {
+        const record = toRecord(previous, (c) => c.chatId);
+
+        const updated = next.filter((c) =>
+            c.chatId !== this.selectedChatId &&
```

Comment:
On reflection, does this selectedChatId check really achieve anything? It's only checking at the point where we enqueue the chats to prime. By the time we actually process the queue, the selected chat could be something different couldn't it. So we could either fail to prime a chat that we _should_ or incorrectly prime one that we don't need to.

Reply:
The reason behind it was that if you're on a chat which then gets updated, the latest events will already get loaded, so if the cache primer also loads the latest events then you'll be double loading them.
However, if you're on a chat but not at the bottom, those new events won't get loaded, which means that if you click the 'go to latest' button it'll need to load the events from the canister.
Because of this I've removed everything to do with the selectedChatId which means clicking the 'go to latest' button will always come from the cache but it does mean we will sometimes be double requesting events.

Reply:
I think sometimes double loading events is a small price to pay for it always being quick.
We can potentially avoid the double loading, I'll have a play around with that next week.